### PR TITLE
Add athenz authentication plugin for C++ client

### PR DESCRIPTION
**Pull request by yush1ga**
### Motivation

Now, we don't have Authentication Plugin of Athenz for C++ client because Athenz client for C++ doesn't exist.

### Modifications
- Add ZTSClient for C++
  - create principal token signed by private key
  - get role token from zts server and cache it
- Add Athenz Plugin for C++
- Add tests for added plugin

### Result

We can use Athenz on C++ client.
Although Athenz supports ECDSA and RSA signing, this ZTS client support only RSA because almost users need only it.

### CODE REVIEW

**Comment by yush1ga** [comment]
I asked Athenz team when Athenz C++ client was released and they answered it was not decided.
Therefore, we don't know when Python client usesr as well as C++ client users are able to use Athenz.

I think following steps are better.

1. Release Athenz plugin for C++ with our original ZTSClient.
2. Replace ZTSClient when Athenz C++ client is released.

I think step 2 is not difficult and maybe we need not change authentication plugin API.

**Comment by saandrews** [comment]
>I asked Athenz team when Athenz C++ client was released and they answered it was not decided.
Therefore, we don't know when Python client usesr as well as C++ client users are able to use Athenz.

>I think following steps are better.

>1. Release Athenz plugin for C++ with our original ZTSClient.
>2. Replace ZTSClient when Athenz C++ client is released.
I think step 2 is not difficult and maybe we need not change authentication plugin API.

@yush1ga We should add the Athens client implementation within Athens repo. It is better to keep Pulsar code clean from adding dependent services used by Pulsar. Can you create a pull request against Athens to get the ZTSClient? When they are ready to open source their current implementation, they can replace/reorganize it.

File: pulsar-client-cpp/lib/ZTSClient.cc [hunk 1]
```diff
@@ -0,0 +1,288 @@
...
+                LOG_ERROR(requiredParams[i] << " parameter is required");
+            }
+        }
+        if (!valid) {
+            LOG_ERROR("Some parameters are missing")
+            return;
+        }
+        
+        // set required value
+        tenantDomain_ = params["tenantDomain"];
```
Comment (rdhabalia) [inline pulsar-client-cpp/lib/ZTSClient.cc]: should we do `requiredParams[0]` in order to avoid any typo ?

**Comment by yush1ga** [comment]
@saandrews 
> Can you create a pull request against Athens to get the ZTSClient? 

I have to add many codes to create a PR against Athenz repo because this ZTSClient has only minimum functions for Pulsar.
I think it is difficult to do it.

**Comment by merlimat** [comment]
@yush1ga @saandrews I think it should be ok to have the simple client code here, as long as we put it in the `athenz.so` plugin. It would not cause any issue to someone not using it.

File: pulsar-client-cpp/lib/ZTSClient.h [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+        std::string privateKeyPath_;
+        std::string ztsUrl_;
+        std::string keyId_;
+        std::string principalHeader_;
+        std::string roleHeader_;
+        int tokenExpirationTime_;
+        static std::map<std::string, RoleToken> roleTokenCache_;
+        static std::string getSalt();
+        static std::string ybase64Encode(const unsigned char *input, int length);
+        const std::string getPrincipalToken() const;
```
Comment (yush1ga) [inline pulsar-client-cpp/lib/ZTSClient.h]: @merlimat 
If we do so, this method may return reference to temporary or local variable.

**Pull request merged**